### ISSUE TITLE
Slow performance rendering chart images

### ISSUE THREAD

**Issue by mattdoran** [wcharczuk/go-chart#28]
Some graphs take a disproportionately long time to render to images.

Sorry I haven't had time to investigate this further yet .... but here's some stuff...

I noticed that there is some really odd variances in the speed of rendering ... even when there is similar number of points and the identical chart rendering code.  The only variables were really the number of series (don't think that was it) ... and the scale/etc of the values.  Odd.

The profiling I was doing showed lots of time spent in Rasterizer.findCell().

This graph plots only about 5000 points at only 360 points/second ...

![applog-xmlrpc-calls](https://cloud.githubusercontent.com/assets/577779/23750845/e3efe91e-0522-11e7-82cd-81f85ca04104.png)

with a profile like this ...

      flat  flat%   sum%        cum   cum%
    19.33s 86.53% 86.53%     19.36s 86.66%  github.com/golang/freetype/raster.(*Rasterizer).findCell
     1.94s  8.68% 95.21%     21.36s 95.61%  github.com/wcharczuk/go-chart/drawing.(*LineStroker).End
     0.65s  2.91% 98.12%      0.65s  2.91%  runtime.usleep
     0.02s  0.09% 98.21%     19.39s 86.79%  github.com/golang/freetype/raster.(*Rasterizer).setCell
     0.01s 0.045% 98.25%     19.42s 86.93%  github.com/golang/freetype/raster.(*Rasterizer).Add1
     0.01s 0.045% 98.30%     19.37s 86.71%  github.com/golang/freetype/raster.(*Rasterizer).saveCell

then another graph using the same code renders ~15,000 points at 139,262 points/second!!

![applog-db-connections](https://cloud.githubusercontent.com/assets/577779/23750877/076f2a58-0523-11e7-80e8-73e48b45929f.png)

with profile ...

      flat  flat%   sum%        cum   cum%
      30ms 33.33% 33.33%       30ms 33.33%  image/png.filter
      20ms 22.22% 55.56%       20ms 22.22%  github.com/golang/freetype/raster.(*RGBAPainter).Paint
      10ms 11.11% 66.67%       10ms 11.11%  compress/flate.(*Writer).Close
      10ms 11.11% 77.78%       20ms 22.22%  compress/flate.(*compressor).deflate
      10ms 11.11% 88.89%       10ms 11.11%  github.com/golang/freetype/raster.(*Rasterizer).findCell

**Comment by wcharczuk** [comment]
Awesome stuff!

Yeah this was my hunch as well, that the drawing calls were the bottleneck and not really anything to do with the chart itself. 

A couple things; the drawing library is compartmentalized enough that we could change how it works pretty easily.

I would be hesitant to use openGL (a.k.a. engi) or similar because they have native requirements. 

What that leaves is maybe digging into the freetype source and figuring out why it's slow and maybe opening pulls against it. 

What do you think?

**Comment by mattdoran** [comment]
Yeah, my spider sense is that something at some layer is doing something very silly.   20 seconds to render this chart is crazy.

Guaranteed there is a way to fix it .... gotta grok what's going on a bit first.  I'm new to go profiling and lack of debugger (well I haven't tried one) ... my time is tight, but will see what I can do.

**Comment by AndyNortrup** [comment]
Matt the gogland ide from jetbrains has a very nice go debugger. Highly recommend.

**Comment by mattdoran** [comment]
@AndyNortrup I'm on their early access, have installed and haven't yet switched from IDEA's plugin ... prob should make the switch. :)
